### PR TITLE
feat: add register and claim logic

### PR DESCRIPTION
**Pull request by edisontim**
## Release Notes

- **New Features**
	- Introduced a comprehensive leaderboard system, allowing users to register contributions and claim rewards.
	- Added functionality for managing leaderboard registration periods and point allocation.
	- Implemented error handling for various leaderboard actions to ensure proper conditions are met.
	- Added a new module for season management, integrating with existing systems.
	- Introduced a `Rewards` component to display leaderboard-related information and facilitate user interactions.

- **Tests**
	- Added a suite of tests to validate leaderboard functionalities, including registration and reward claiming scenarios.
	- Expanded testing coverage for edge cases related to leaderboard actions.

- **Documentation**
	- Enhanced namespace definitions to include new models and systems related to seasons and leaderboards.
	- Updated testing workflows to include new season-related tests.

### CODE REVIEW

File: contracts/src/models/season.cairo [hunk 1]
```diff
@@ -39,3 +39,60 @@ pub impl SeasonImpl of SeasonTrait {
...
+}
+
+#[generate_trait]
+pub impl LeaderboardEntryCustomImpl of LeaderboardEntryCustomTrait {
+    fn get(ref world: WorldStorage, address: starknet::ContractAddress) -> LeaderboardEntry {
+        let entry: LeaderboardEntry = world.read_model(address);
+        entry
+    }
+
+    fn append(ref self: Leaderboard, ref world: WorldStorage, address: starknet::ContractAddress, points: u128) {
```
Comment (aymericdelab) [inline contracts/src/models/season.cairo:85]: i would call this register, to reflect more what it actually does

File: contracts/src/systems/hyperstructure/contracts.cairo [hunk 2]
```diff
@@ -8,6 +8,8 @@ use s0_eternum::{
 };
 use starknet::ContractAddress;
 
+const LEADERBOARD_REGISTRATION_PERIOD: u64 = 604800; // one week
```
Comment (aymericdelab) [inline contracts/src/systems/hyperstructure/contracts.cairo:11]: can we put this into a config ?

File: contracts/src/systems/hyperstructure/contracts.cairo [hunk 3]
```diff
@@ -345,6 +347,10 @@ mod hyperstructure_systems {
             let winner_address = starknet::get_caller_address();
             world.emit_event(@GameEnded { winner_address, timestamp: starknet::get_block_timestamp() });
 
+            let mut leaderboard: Leaderboard = world.read_model(WORLD_CONFIG_ID);
+            leaderboard.registration_end_timestamp = starknet::get_block_timestamp() + LEADERBOARD_REGISTRATION_PERIOD;
```
Comment (aymericdelab) [inline contracts/src/systems/hyperstructure/contracts.cairo:351]: do we need to store registration end, if we already have GameEnded timestamp and LEADERBOARD_REGISTRATION_PERIOD?

File: contracts/src/systems/hyperstructure/contracts.cairo [hunk 4]
```diff
@@ -561,15 +567,15 @@ mod hyperstructure_systems {
...
         }
 
         fn get_total_points_percentage(
             total_contributable_amount: u128, resource_rarity: u128, resource_quantity: u128,
         ) -> u128 {
             // resource rarity already has a x100 factor in
-            let effective_contribution = (resource_quantity / RESOURCE_PRECISION) * (resource_rarity);
-            effective_contribution * 1_000_000 / total_contributable_amount
+            let effective_contribution = (resource_quantity * resource_rarity) / RESOURCE_PRECISION;
+            (effective_contribution * 1_000_000) / total_contributable_amount
```
Comment (aymericdelab) [inline contracts/src/systems/hyperstructure/contracts.cairo:578]: is this leaderboard precision ? maybe create variable to explain what it is

File: contracts/src/systems/hyperstructure/tests.cairo [hunk 5]
```diff
@@ -343,6 +343,29 @@ fn hyperstructure_test_end_game_failure_completion_and_shares() {
...
+    starknet::testing::set_account_contract_address(contract_address_const::<'player1'>());
+
+    let hyperstructure_entity_id_0 = spawn_and_finish_hyperstructure(
+        ref world, hyperstructure_systems_dispatcher, realm_entity_id, Coord { x: 0, y: 0 }
+    );
+
+    hyperstructure_systems_dispatcher
+        .set_co_owners(hyperstructure_entity_id_0, array![(contract_address_const::<'player1'>(), 10_000)].span());
+
+    starknet::testing::set_block_timestamp(1001);
```
Comment (aymericdelab) [inline contracts/src/systems/hyperstructure/tests.cairo:362]: if 1_000 points/second

2_000_000 points for completion and 1_000 * 1_001 points at timestamp 1001, then total point for player is 3_000_000 no? why does this fail

File: contracts/src/systems/season/contracts.cairo [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            resources::contracts::resource_bridge_systems::{ERC20ABIDispatcher, ERC20ABIDispatcherTrait}
+        },
+    };
+    use starknet::ContractAddress;
+
+    pub const SCALING_FACTOR: u256 = 1_000_000;
+
+    #[abi(embed_v0)]
+    impl SeasonSystemsImpl of super::ISeasonSystems<ContractState> {
+        fn register_to_leaderboard(
```
Comment (aymericdelab) [inline contracts/src/systems/season/contracts.cairo:40]: check if registration period has started

File: contracts/src/systems/season/contracts.cairo [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                    ref world, hyperstructures_contributed_to
+                );
+
+            total_points +=
+                InternalHyperstructureSystemsImpl::compute_total_share_points(world, hyperstructure_shareholder_epochs);
+
+            leaderboard.append(ref world, starknet::get_caller_address(), total_points);
+        }
+
+        fn claim_leaderboard_rewards(ref self: ContractState, token: ContractAddress) {
```
Comment (aymericdelab) [inline contracts/src/systems/season/contracts.cairo:65]: add the fact that registration period needs to be ended

File: contracts/src/systems/season/tests.cairo [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+
+    starknet::testing::set_contract_address(contract_address_const::<'player1'>());
+
+    season_systems_dispatcher.claim_leaderboard_rewards(mock_erc20_address);
+    season_systems_dispatcher.claim_leaderboard_rewards(mock_erc20_address);
+}
+
+#[test]
+#[should_panic(expected: ("Registration period is over", 'ENTRYPOINT_FAILED'))]
+fn season_test_register_too_late() {
```
Comment (aymericdelab) [inline contracts/src/systems/season/tests.cairo:257]: add test register too early

File: contracts/src/systems/hyperstructure/contracts.cairo (hunk 2 above)
Comment (edisontim) [reply contracts/src/systems/hyperstructure/contracts.cairo:11]: should be hardcoded in the contracts so we can't change it, loaf mentioned it should be as permissionless as possible

File: contracts/src/systems/hyperstructure/contracts.cairo (hunk 3 above)
Comment (edisontim) [reply contracts/src/systems/hyperstructure/contracts.cairo:351]: the `GameEnded` is an event being emitted, there's no way to get it from the contracts. We would have to switch it to a model then

File: contracts/src/systems/hyperstructure/tests.cairo (hunk 5 above)
Comment (edisontim) [reply contracts/src/systems/hyperstructure/tests.cairo:362]: we set the caller's address to `player2` but `player1` has all the points, this was to test that another caller wouldn't have the points of someone else

File: contracts/src/systems/season/contracts.cairo (hunk 6 above)
Comment (edisontim) [reply contracts/src/systems/season/contracts.cairo:40]: This should fail anyways as if the season isn't over the leaderboard wouldn't have been set before so `registration_end_timestamp` would be 0 so this line would panic 
```
            assert!(
                leaderboard.registration_end_timestamp > starknet::get_block_timestamp(), "Registration period is over"
            );
```
do you think it should be more explicit?

File: contracts/src/systems/hyperstructure/contracts.cairo (hunk 2 above)
Comment (aymericdelab) [reply contracts/src/systems/hyperstructure/contracts.cairo:11]: ok

File: contracts/src/systems/hyperstructure/contracts.cairo (hunk 3 above)
Comment (aymericdelab) [reply contracts/src/systems/hyperstructure/contracts.cairo:353]: ok

File: contracts/src/systems/hyperstructure/tests.cairo (hunk 5 above)
Comment (aymericdelab) [reply contracts/src/systems/hyperstructure/tests.cairo:362]: ok

File: contracts/src/systems/season/contracts.cairo (hunk 6 above)
Comment (edisontim) [reply contracts/src/systems/season/contracts.cairo:40]: yeah just created a test for it
